### PR TITLE
2d link check, EdgeCollapse in progress

### PR DESCRIPTION
**Pull request by yfzhou42**

### CODE REVIEW

File: app/tests/test_2d_operations.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#include <iostream>
+#include "EdgeCollapse/EdgeCollapse.h"
+
+using namespace wmtk;
+
+// these roll back tests must have more than 2 triangles, since by design when there are only two
+// incident triangles the edge cannot be collapsed
+// this check should be doen in pre collapse check
+
+TEST_CASE("edge_collapse", "[test_2d_operation]")
```
Comment (teseoch) [inline app/tests/test_2d_operations.cpp:13]: this doesn't depend on the app, if true, should go in the tri mesh tests

File: app/tests/test_2d_operations.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    TriMesh::Tuple dummy;
+
+    bool collapse = m.collapse_edge(tuple, dummy);
+
+    REQUIRE(collapse);
+    REQUIRE_FALSE(tuple.is_valid(m));
+    REQUIRE(m.check_mesh_connectivity_validity());
+}
+
+TEST_CASE("rollback_collapse_operation", "[test_2d_operation]")
```
Comment (teseoch) [inline app/tests/test_2d_operations.cpp:29]: this doesn't depend on the app, if true, should go in the tri mesh tests

File: src/wmtk/TriMesh.h [hunk 3]
```diff
@@ -311,7 +310,22 @@ class TriMesh
...
+    {
+        auto v1_conn_tris = m_vertex_connectivity[t.get_vid()].m_conn_tris;
+        auto v2_conn_tris = m_vertex_connectivity[switch_vertex(t).get_vid()].m_conn_tris;
+        size_t fid1 = t.get_fid();
+        Tuple tmp_tuple = switch_face(t).value_or(t);
+        size_t fid2 = tmp_tuple.get_fid();
+        vector_erase(v1_conn_tris, fid1);
+        vector_erase(v1_conn_tris, fid2);
+        vector_erase(v2_conn_tris, fid1);
+        vector_erase(v2_conn_tris, fid2);
```
Comment (teseoch) [inline src/wmtk/TriMesh.h:324]: what is all of this?

Comment (yfzhou42) [reply src/wmtk/TriMesh.h:324]: I set the constraint that if all the triangles left depend on the edge to be collapsed then this edge cannot be collapsed. Basically, after all the collapsing there has to be at least one triangle left

File: app/EdgeCollapse/EdgeCollapse.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    ElementInQueue(const wmtk::TriMesh::Tuple& e, double w)
+        : edge(e)
+        , weight(w)
+    {}
+};
+struct cmp_l
+{
+    bool operator()(const ElementInQueue& e1, const ElementInQueue& e2)
+    {
+        if (e1.weight == e2.weight) return e1.edge.get_vid() > e2.edge.get_vid();
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/EdgeCollapse.h:49]: This is not really necessary.

File: app/EdgeCollapse/ShortestEdgeCollapse.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    if (new_vid < m_vertex_positions.size()) m_vertex_positions[new_vid] = new_position;
+    m_vertex_positions.emplace_back(new_position);
+}
+
+
+bool Edge2d::EdgeCollapse::collapse_shortest()
+{
+    std::vector<TriMesh::Tuple> edges = get_edges();
+    std::priority_queue<ElementInQueue, std::vector<ElementInQueue>, cmp_s> ec_queue;
+    double shortest = DBL_MAX;
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/ShortestEdgeCollapse.cpp:23]:
```suggestion
    double shortest = std::numerical_limits<double>::max();
```

File: app/EdgeCollapse/ShortestEdgeCollapse.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#include <float.h>
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/ShortestEdgeCollapse.cpp:1]:
```suggestion
```

File: src/wmtk/TriMesh.h (hunk 3 above)
Comment (jiangzhongshi) [reply src/wmtk/TriMesh.h:324]: I see now. This seems to be equivalent to, and maybe not necesary.
```
if (v1_conn_tris.size() == 2 && v2_conn_tris.size() == 2 )
```

File: src/wmtk/TriMesh.h [hunk 7]
```diff
@@ -311,7 +310,22 @@ class TriMesh
...
+
+        if (check_link(t) && (v1_conn_tris.size() + v2_conn_tris.size() > 0)) return true;
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.h:326]:
```suggestion
        if (check_link(t) && (v1_conn_tris.size() + v2_conn_tris.size() > 4)) return true;
```

File: app/EdgeCollapse/QEC.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
+#include <wmtk/TriMesh.h>
+#include <wmtk/utils/VectorUtils.h>
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include "EdgeCollapse.h"
+void add_Qs(std::array<double, 10> Q1, std::array<double, 10> Q2)
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/QEC.cpp:6]: This is a bug

Comment (jiangzhongshi) [inline app/EdgeCollapse/QEC.cpp:6]:
```suggestion
void add_Qs(std::array<double, 10>& Q1, const std::array<double, 10>& Q2)
```

**Comment by jiangzhongshi** [comment]
Also, in the test, verify this is the same as libigl version. (Check same number of vertices in the end)

File: src/wmtk/TriMesh.cpp [hunk 9]
```diff
@@ -117,10 +117,66 @@ std::optional<TriMesh::Tuple> TriMesh::Tuple::switch_face(const TriMesh& m) cons
...
+        if (m_vertex_connectivity[i].m_is_removed) continue;
+        assert(
+            m_vertex_connectivity[i].m_conn_tris == conn_tris[i] &&
+            "m_vertex_connectivity[i].m_conn_tris!=conn_tris[i]");
+    }
+    return true;
+}
+
+// link check, prerequisite for edge collapse
+bool wmtk::TriMesh::check_link(const Tuple& edge) const
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.cpp:141]: rename to `check_link_condition`

Comment (jiangzhongshi) [reply src/wmtk/TriMesh.cpp:141]: This can be tricky to make right, add a test for link success or fail.

File: src/wmtk/TriMesh.cpp [hunk 10]
```diff
@@ -263,3 +317,47 @@ void TriMesh::swap_edge(const Tuple& t, int type)
...
+        if (tmp_tuple.get_fid() < current_t.get_fid()) fid = tmp_tuple.get_fid();
+        int j = m_tri_connectivity[fid].find(current_t.get_vid());
+        one_ring_edges.emplace_back(current_t.get_vid(), (j + 2) % 3, fid, *this);
+        tmp_tuple = switch_edge(current_t);
+        current_t = tmp_tuple;
+    }
+    return one_ring_edges;
+};
+
+std::vector<wmtk::TriMesh::Tuple> TriMesh::get_incident_verts_for_tri(
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.cpp:352]: rename to `get_oriented_verts_for_tri`, see the corresponding function in `TetMesh.h`

**Review by jiangzhongshi — CHANGES_REQUESTED** [review]

File: src/wmtk/TriMesh.h (hunk 3 above)
Comment (yfzhou42) [reply src/wmtk/TriMesh.h:324]: it's equivalent to if (v1_conn_tris.size() == 2 || v2_conn_tris.size() == 2 )

File: app/EdgeCollapse/QEC.cpp [hunk 11]
```diff
@@ -3,7 +3,7 @@
 #include <Eigen/Core>
 #include <Eigen/Geometry>
 #include "EdgeCollapse.h"
-void add_Qs(std::array<double, 10> Q1, std::array<double, 10> Q2)
+void add_Qs(std::array<double, 10>& Q1, std::array<double, 10>& Q2)
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/QEC.cpp:6]: add a const for the second parameter, just to be clear.

File: app/EdgeCollapse/ShortestEdgeCollapse.cpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+#include <wmtk/utils/VectorUtils.h>
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include "EdgeCollapse.h"
+
+using namespace Edge2d;
+// delete the old two vert position, add the new at the middle of the old two points
+void Edge2d::EdgeCollapse::update_position(size_t v1, size_t v2, Tuple& new_vert)
+{
+    size_t new_vid = new_vert.get_vid();
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/ShortestEdgeCollapse.cpp:11]: you can directly use `std::min(v1,v2)` as the new_vid. This avoid resizing the array or traversing the array all the time.

File: app/EdgeCollapse/ShortestEdgeCollapse.cpp [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    Eigen::Vector3d new_position = (m_vertex_positions[v1] + m_vertex_positions[v2]) / 2;
+    if (new_vid < m_vertex_positions.size()) m_vertex_positions[new_vid] = new_position;
+    m_vertex_positions.emplace_back(new_position);
```
Comment (jiangzhongshi) [inline app/EdgeCollapse/ShortestEdgeCollapse.cpp:14]: Shouldn't this be an `if-else`?

File: app/tests/test_2d_operations.cpp [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    v_positions[2] = Eigen::Vector3d(3, 3, 0);
+    v_positions[3] = Eigen::Vector3d(0, 0, 0);
+    v_positions[4] = Eigen::Vector3d(0.5, 0, 0);
+    v_positions[5] = Eigen::Vector3d(0, -3, 0);
+    EdgeCollapse m(v_positions);
+    std::vector<std::array<size_t, 3>> tris = {{{0, 1, 3}}, {{1, 2, 4}}, {{3, 1, 4}}, {{3, 4, 5}}};
+    m.create_mesh(6, tris);
+    std::vector<TriMesh::Tuple> edges = m.get_edges();
+    // find the shortest edge
+    double shortest = 100;
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:32]:
```suggestion
    double shortest = std::numerical_limits<double>::max();
```

File: app/tests/test_2d_operations.cpp [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        size_t v1 = t.get_vid();
+        size_t v2 = m.switch_vertex(t).get_vid();
+        if ((v_positions[v1] - v_positions[v2]).squaredNorm() < shortest) {
+            shortest = (v_positions[v1] - v_positions[v2]).squaredNorm();
+            shortest_edge = t;
+        }
+    }
+    m.collapse_shortest();
+    // the collapsed edge tuple is not valid anymore
+    REQUIRE_FALSE(shortest_edge.is_valid(m));
```
Comment (jiangzhongshi) [inline app/tests/test_2d_operations.cpp:44]: This is not a valid test: the edge tuple will be invalid even itself is not collapsed. You can test that it became your figure drawn above. number of verts = 4, number of tets = 2.

File: src/wmtk/TriMesh.h (hunk 3 above)
Comment (jiangzhongshi) [reply src/wmtk/TriMesh.h:325]: Then why not just write that?

File: app/EdgeCollapse/QEC.cpp (hunk 11 above)
Comment (jiangzhongshi) [reply app/EdgeCollapse/QEC.cpp:6]:
```suggestion
void add_Qs(std::array<double, 10>& Q1, const std::array<double, 10>& Q2)
```

File: app/tests/test_2d_operations.cpp (hunk 15 above)
Comment (jiangzhongshi) [reply app/tests/test_2d_operations.cpp:44]:
```
REQUIRE(valid_v == 4);
REQUIRE(valid_t == 2);
```
you can just count them.

File: src/wmtk/TriMesh.cpp [hunk 16]
```diff
@@ -117,11 +117,69 @@ std::optional<TriMesh::Tuple> TriMesh::Tuple::switch_face(const TriMesh& m) cons
...
+        if (m_vertex_connectivity[i].m_is_removed) continue;
+        assert(
+            m_vertex_connectivity[i].m_conn_tris == conn_tris[i] &&
+            "m_vertex_connectivity[i].m_conn_tris!=conn_tris[i]");
+    }
+    return true;
+}
+
+// link check, prerequisite for edge collapse
+bool wmtk::TriMesh::check_link_condition(const Tuple& edge) const
```
Comment (jiangzhongshi) [inline src/wmtk/TriMesh.cpp:141]: link condition should have its own test.

**Review by jiangzhongshi — DISMISSED** [review]

File: app/EdgeCollapse/QEC.cpp (hunk 11 above)
Comment (jiangzhongshi) [reply app/EdgeCollapse/QEC.cpp:6]: Also, please add a test on a model, where the result is the same as libigl.

File: src/wmtk/TriMesh.cpp (hunk 16 above)
Comment (yfzhou42) [reply src/wmtk/TriMesh.cpp:141]: There is one

Comment (jiangzhongshi) [reply src/wmtk/TriMesh.cpp:141]: Thanks, I see it now.

File: app/tests/test_2d_operations.cpp [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+TEST_CASE("shortest_edge_collapse_on_mesh", "[test_2d_operations]")
+{
+    // const std::string root(WMT_DATA_DIR);
+    // const std::string path = root + "/Octocat.obj";
+    const std::string path = "/Users/yunfanzhou/Downloads/tmp/initial.obj";
```
Comment (teseoch) [inline app/tests/test_2d_operations.cpp:49]: don't use absolute paths

**Comment by jdumas** [comment]
Hi. Please be careful, do add `app/data` as a submodule!

Screen Shot 2022-01-04 at 10 42 59 PM

File: app/tests/test_2d_operations.cpp [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        size_t v2 = m.switch_vertex(t).get_vid();
+        if ((v_positions[v1] - v_positions[v2]).squaredNorm() < shortest) {
+            shortest = (v_positions[v1] - v_positions[v2]).squaredNorm();
+            shortest_edge = t;
+        }
+    }
+    m.collapse_shortest(100);
+    // the collapsed edge tuple is not valid anymore
+    REQUIRE_FALSE(shortest_edge.is_valid(m));
+    m.write_triangle_mesh("/Users/yunfanzhou/Downloads/tmp/collapsed.obj");
```
Comment (teseoch) [inline app/tests/test_2d_operations.cpp:77]: Remove this

File: src/wmtk/TriMesh.h [hunk 19]
```diff
@@ -300,38 +299,109 @@ class TriMesh
     size_t get_next_empty_slot_t()
     {
         m_tri_connectivity.emplace_back();
```
Comment (teseoch) [inline src/wmtk/TriMesh.h:301]: Do we need this in 3d @Yixin-Hu

File: src/wmtk/TriMesh.h [hunk 20]
```diff
@@ -300,38 +299,109 @@ class TriMesh
...
+        if (check_link_condition(t) && (v1_conn_tris.size() + v2_conn_tris.size() > 0)) return true;
+        return false;
+    }
+    virtual bool collapse_after(const Tuple& t)
+    {
+        if (check_mesh_connectivity_validity() && t.is_valid(*this)) return true;
+        return false;
+    }
+
+    virtual void resize_attributes(size_t v, size_t t) {}
```
Comment (teseoch) [inline src/wmtk/TriMesh.h:347]: Add e, right?

File: src/wmtk/TriMesh.h (hunk 19 above)
Comment (teseoch) [reply src/wmtk/TriMesh.h:301]: Also we need to resize edge attributes

File: src/wmtk/TriMesh.h [hunk 21]
```diff
@@ -300,38 +299,109 @@ class TriMesh
...
 
 public:
     size_t n_triangles() const { return m_tri_connectivity.size(); }
     size_t n_vertices() const { return m_vertex_connectivity.size(); }
 
+    // just for testing
+    std::vector<TriangleConnectivity> get_m_tri_connectivity() const { return m_tri_connectivity; }
```
Comment (teseoch) [inline src/wmtk/TriMesh.h:354]: We discussed that we cant test private stuff :(

**Comment by teseoch** [comment]
> Hi. Please be careful, do add `app/data` as a submodule!
> 
> Screen Shot 2022-01-04 at 10 42 59 PM

What do you mean? We use the cmake stuff to get the data

**Comment by jdumas** [comment]
> What do you mean? We use the cmake stuff to get the data

Exactly, so the folder `app/data` shouldn't be committed as a git submodule!

File: app/EdgeCollapse/EdgeCollapse.h [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        Eigen::Vector3d v1p;
+        Eigen::Vector3d v2p;
+    } collapse_cache;
+
+
+    bool collapse_before(const Tuple& t) override;
+    bool collapse_after(const Tuple& t) override;
+
+    void compact_vertex_positions(std::vector<size_t> vid_newvid_map)
```
Comment (teseoch) [inline app/EdgeCollapse/EdgeCollapse.h:33]: Is this necessary?

File: app/EdgeCollapse/EdgeCollapse.h [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+        }
+        new_m_vertex_positions.resize(cnt);
+        m_vertex_positions = new_m_vertex_positions;
+    };
+
+    // write the collapsed mesh into a obj
+    void write_triangle_mesh(std::string path)
+    {
+        consolidate_mesh_connectivity();
+        std::vector<VertexConnectivity> new_m_vertex_connectivity = get_m_vertex_connectivity();
```
Comment (teseoch) [inline app/EdgeCollapse/EdgeCollapse.h:50]: Why not calling the other method?

File: src/wmtk/TriMesh.cpp [hunk 24]
```diff
@@ -263,3 +342,194 @@ void TriMesh::swap_edge(const Tuple& t, int type)
...
+    f = 0;
+    while (m_vertex_connectivity[f].m_is_removed == false) ++f;
+    m_vertex_connectivity.resize(f);
+
+    f = 0;
+    while (m_tri_connectivity[f].m_is_removed == false) ++f;
+    m_tri_connectivity.resize(f);
+
+    // Resize user class attributes
+    resize_attributes(m_vertex_connectivity.size(), m_tri_connectivity.size());
```
Comment (teseoch) [inline src/wmtk/TriMesh.cpp:469]: No edge attributes?

File: src/wmtk/TriMesh.h [hunk 25]
```diff
@@ -300,38 +300,113 @@ class TriMesh
...
 
 protected:
     virtual bool split_before(const Tuple& t) { return true; }
     virtual bool split_after(const Tuple& t) { return true; }
-    virtual bool collapse_before(const Tuple& t) { return true; }
-    virtual bool collapse_after(const Tuple& t) { return true; }
+    // check link, check if it's the last edge
+    virtual bool collapse_before(const Tuple& t)
+    {
+        if (check_link_condition(t) && check_manifold(t)) return true;
```
Comment (teseoch) [inline src/wmtk/TriMesh.h:325]: remove `check_manifold` change `check_link_condition` to include open boundaries
test `check_link_condition` vs libigl

File: src/wmtk/TriMesh.h [hunk 26]
```diff
@@ -300,38 +300,113 @@ class TriMesh
...
+        return false;
+    }
+    virtual bool collapse_after(const Tuple& t)
+    {
+        if (check_mesh_connectivity_validity() && t.is_valid(*this)) return true;
```
Comment (teseoch) [inline src/wmtk/TriMesh.h:330]: these should be assertions

**Pull request merged**